src/filters/filters.jl

```diff
@@ -106,8 +234,75 @@ function Base.append!(filters::FilterPipeline, extra)
     end
     return filters
 end
+function Base.push!(p::FilterPipeline, f::F) where F <: Filter
+    ref = Ref(f)
+    GC.@preserve ref begin
+        API.h5p_set_filter(p.plist, filterid(F), API.H5Z_FLAG_OPTIONAL, div(sizeof(F), sizeof(Cuint)), pointer_from_objref(ref))
+    end
+    return p
+end
+function Base.push!(p::FilterPipeline, f::UnknownFilter)
+    GC.@preserve f begin
+        API.h5p_set_filter(p.plist, f.filter_id, f.flags, length(f.data), pointer(f.data))
+    end
+end
+
+# Generic implementation of register_filter
+"""
+    register_filter(filter::F) where F <: Filter
+
+Register the filter with the HDF5 library via API.h5z_register.
+Also add F to the FILTERS dictionary.
+"""
+function register_filter(filter::F) where F <: Filter
+    id = filterid(filter)
+    encoder = encoder_present(filter)
+    decoder = decoder_present(filter)
+    name = filtername(filter)
+    can_apply = can_apply_cfunc(filter)
+    set_local = set_local_cfunc(filter)
+    func = filter_cfunc(filter)
+    GC.@preserve name begin
+        API.h5z_register(API.H5Z_class_t(
+            API.H5Z_CLASS_T_VERS,
+            id,
+            encoder,
+            decoder,
+            pointer(name),
+            can_apply,
+            set_local,
+            func
+        ))
+    end
+    # Should this be the filter instance rather than the type?
+    FILTERS[id] = F
+    return nothing
+end
+register_filter(::Type{F}) where {F<:Filter} = register_filter(F())
```

Comment:
This presumes that `F()` is defined.

Reply:
The idea is that individual filters may override this, but this is the default implementation.

Reply:
For example, for Bzip2, this is just
```julia
register_filter(::Type{Bzip2Filter}) = register_bzip2()
register_filter(::Bzip2Filter) = register_bzip2()
```

At the moment, none of the new filters I have implemented use the generic `register_filter`. I may just remove this in the end.

Reply:
I just left a method stub to retain the documentation in 7303742

Reply:
Could you explain why we have both `register_filter` defined on the type and instance. Isn't that just redundant? Why wouldn't we just use one of the definitions. 

I'm not sure if the instance or type definition is better, but it looks like just using one of the definitions would be simpler and sufficient. Note, if the method defined on the type needs to be deprecated that's also fine.